### PR TITLE
fix(contracts): allow template initialization directly after deployment

### PR DESCRIPTION
**Pull request by clemsos**
# Description

This allows to initialise PublicLock directly after deploying the contract or during the setting of the template in Unlock. This will also allow to reuse a template - before the tx will revert when trying to set a template that has already been set.

# Issues

Fixes #
Refs #

# Checklist:

- [ ] 1 PR, 1 purpose: my Pull Request applies to a single purpose
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the docs to reflect my changes if applicable
- [ ] I have added tests (and stories for frontend components) that prove my fix is effective or that my feature works
- [ ] I have performed a self-review of my own code
- [ ] If my code involves visual changes, I am adding applicable screenshots to this thread

### CODE REVIEW

**Comment by clemsos** [comment]
The Unlock contract needs to be re-packaged before merging but better to do it after #13197 is merged

File: smart-contracts/contracts/Unlock.sol [hunk 1]
```diff
@@ -195,10 +195,14 @@ contract Unlock is UnlockInitializable, UnlockOwnable {
    * @dev This will initialize the template and revokeOwnership.
    */
   function addLockTemplate(address impl, uint16 version) public onlyOwner {
-    // First claim the template so that no-one else could
-    // this will revert if the template was already initialized.
-    IPublicLock(impl).initialize(address(this), 0, address(0), 0, 0, "");
-    IPublicLock(impl).renounceLockManager();
+    // if the template has not been initialized,
+    // claim the template so that no-one else could
+    try IPublicLock(impl).initialize(address(this), 0, address(0), 0, 0, "") {
```
Comment (julien51) [inline smart-contracts/contracts/Unlock.sol:200]: 👍 Great idea to use try/catch here!

File: smart-contracts/contracts/Unlock.sol [hunk 2]
```diff
@@ -195,10 +195,14 @@ contract Unlock is UnlockInitializable, UnlockOwnable {
...
+      // renounce the lock manager role that was added during initialization
+      IPublicLock(impl).renounceLockManager();
+    } catch {
+      // failure means that the template is already initialized
```
Comment (julien51) [inline smart-contracts/contracts/Unlock.sol:204]: Would there be an easy way to verify that the contract does not have any lock nanager? Probably not?
That would be a check we can do here.

**Review by julien51 — APPROVED** [review]

File: smart-contracts/contracts/Unlock.sol (hunk 2 above)
Comment (clemsos) [reply smart-contracts/contracts/Unlock.sol:204]: No, there is no way to list the lock managers atm

Comment (julien51) [reply smart-contracts/contracts/Unlock.sol:204]: ok, thanks!

**Pull request merged**